### PR TITLE
Separate Abstract registry and AbstractStyxService from each other 

### PR DESCRIPTION
**Pull request by mikkokar**

### CODE REVIEW

File: components/proxy/src/main/java/com/hotels/styx/routing/handlers/BackendServiceProxy.java [hunk 1]
```diff
@@ -91,18 +90,12 @@ public HttpHandler2 build(List<String> parents, RouteHandlerFactory x, RouteHand
             String provider = config.get("backendProvider")
                     .orElseThrow(() -> missingAttributeError(configBlock, join(".", parents), "backendProvider"));
 
-            StyxService service = services.get(provider);
-            if (service == null) {
+            Registry<BackendService> registry = backendRegistries.get(provider);
+            if (registry == null) {
```
Comment (kvosper) [inline components/proxy/src/main/java/com/hotels/styx/routing/handlers/BackendServiceProxy.java]: Looks like we are reinventing `Preconditions.checkNotNull` here.

File: components/proxy/src/test/java/com/hotels/styx/infrastructure/AbstractRegistryTest.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        AbstractRegistry.ChangeListener<IdObject> listener1 = mock(AbstractRegistry.ChangeListener.class);
+        registry.addListener(listener1);
+        verify(listener1).onChange(any(Registry.Changes.class));
+
+        registry.set(ImmutableList.of(idObject("a", "1"), idObject("b", "2")));
+        verify(listener1, times(2)).onChange(any(Registry.Changes.class));
+
+        // Does not change:
+        registry.set(ImmutableList.of(idObject("a", "1"), idObject("b", "2")));
+        verify(listener1, times(2)).onChange(any(Registry.Changes.class));
```
Comment (kvosper) [inline components/proxy/src/test/java/com/hotels/styx/infrastructure/AbstractRegistryTest.java]: Perhaps it would be more intuitive to use `verifyNoMoreInteractions(listener1);`

File: components/proxy/src/test/java/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistryTest.java [hunk 3]
```diff
@@ -15,232 +15,142 @@
...
-        registry.addListener(changeListener);
-        StyxFutures.await(registry.start());
+        verify(delegate).reload();
     }
 
     @Test
-    public void firesChangeEventForInitialState() {
-        verify(changeListener).onChange(anyChanges());
-    }
+    public void relaysAddListenerToRegisrtryDelegate() {
```
Comment (kvosper) [inline components/proxy/src/test/java/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistryTest.java]: Typo

File: components/proxy/src/test/java/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistryTest.java [hunk 4]
```diff
@@ -15,232 +15,142 @@
...
+        verify(delegate).addListener(eq(listener));
     }
 
     @Test
-    public void discardsInvalidHealthCheckURIs2() throws Throwable {
-        registry = new FileBackedBackendServicesRegistry(givenOrigins("/backends/origins-with-invalid-healthcheck-uri.yml"));
-        registry.addListener(changeListener);
-        unwrapException(() -> registry.start());
-    }
+    public void relaysRemoveListenerToRegisrtryDelegate() {
```
Comment (kvosper) [inline components/proxy/src/test/java/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistryTest.java]: Typo

File: components/proxy/src/test/java/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistryTest.java [hunk 5]
```diff
@@ -15,232 +15,142 @@
...
     }
 
     @Test
-    public void firesChangeEventWhenBackendOriginHostsAreUpdated() throws IOException, ExecutionException, InterruptedException {
-        givenOrigins("/backends/origins-host-changed.yml");
-        ReloadResult result = reload().get();
-        assertThat(result.outcome(), is(RELOADED));
-        verify(changeListener).onChange(updated("webapp"));
-    }
+    public void relaysGetToRgistryDelegate() {
```
Comment (kvosper) [inline components/proxy/src/test/java/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistryTest.java]: Typo

File: components/proxy/src/test/java/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistryTest.java [hunk 6]
```diff
@@ -15,232 +15,142 @@
...
     }
 
     @Test
-    public void firesChangeEventWhenBackendIsAdded() throws IOException, ExecutionException, InterruptedException {
-        givenOrigins("/backends/origins-pd-app-added.yml");
-        ReloadResult result = reload().get();
-        assertThat(result.outcome(), is(RELOADED));
-        verify(changeListener).onChange(added("product-details"));
-    }
+    public void passesThroughUnchangedReloadResult() {
```
Comment (kvosper) [inline components/proxy/src/test/java/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistryTest.java]: The name seems to suggest it's going to verify the output of the reload, but it doesn't. What is the objective here?

File: components/proxy/src/test/java/com/hotels/styx/serviceproviders/ServiceProvisionTest.java [hunk 7]
```diff
@@ -86,6 +112,15 @@ public void servicesReturnCorrectlyFromCall() {
         )));
     }
 
+    @Test
+    public void isInstanceWorks() {
+        Environment env = environmentWithConfig(yamlForServices);
+        Map<String, StyxService> services = loadServices(env.configuration(), env, "multi", StyxService.class);
+
+        assertThat(services.get("backendProvider"), Matchers.instanceOf(BackendServiceProvider.class));
```
Comment (kvosper) [inline components/proxy/src/test/java/com/hotels/styx/serviceproviders/ServiceProvisionTest.java]: Please use static-imports!

File: components/proxy/src/test/scala/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistrySpec.scala [hunk 8]
```diff
@@ -190,14 +193,14 @@ class FileBackedBackendServicesRegistrySpec extends FunSpec with Eventually {
         registry.reload()
 
         assertThat(registry.get(), contains(genBackendOne))
-        verifyNoMoreInteractions(listener)
```
Comment (kvosper) [inline components/proxy/src/test/scala/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistrySpec.scala]: What is the purpose of this change?

File: components/proxy/src/main/java/com/hotels/styx/infrastructure/AbstractRegistry.java [hunk 9]
```diff
@@ -39,39 +38,43 @@
...
 
     @Override
     public Iterable<T> get() {
         return snapshot.get();
     }
 
     protected void notifyListeners(Changes<T> changes) {
-        LOG.info("notifying about services={} to listeners={}", announcer.listeners());
-        announcer.announce().onChange(changes);
+        if (!changes.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (mikkokar) [reply components/proxy/src/main/java/com/hotels/styx/infrastructure/AbstractRegistry.java]: Hmm... I guess that would make sense.

File: components/proxy/src/test/java/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistryTest.java (hunk 6 above)
Comment (mikkokar) [reply components/proxy/src/test/java/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistryTest.java]: I guess this can be removed.

File: components/proxy/src/test/scala/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistrySpec.scala (hunk 8 above)
Comment (mikkokar) [reply components/proxy/src/test/scala/com/hotels/styx/proxy/backends/file/FileBackedBackendServicesRegistrySpec.scala]: To prevent test from failing. `verifyNoMoreInteractions` appears to be picking up some uninteresting interactions from earlier on in the tests, which we did not verify for in this test.

**Comment by mikkokar** [comment]
I have accidentally committed some future work into this branch, which needs to be backed off. For this reason I will:
- Close this PR as it is.
- Create a new branch with relevant changes, rebase that to master, and address all PR findings.
- Publish the new branch in a separate PR.
